### PR TITLE
[ISSUE #801]Rocketmq connector sink for hudi

### PR DESCRIPTION
**Pull request by lizhiboo**
Support rocketmq connector sink to hudi.
welcome to contribute together.

### CODE REVIEW

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/config/CloneUtils.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.rocketmq.connect.hudi.config;
+
+import java.io.*;
```
Comment (ltamber) [inline rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/config/CloneUtils.java:19]: import * is not recommended.

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+import org.apache.rocketmq.common.protocol.route.BrokerData;
+import org.apache.rocketmq.common.protocol.route.QueueData;
+import org.apache.rocketmq.common.protocol.route.TopicRouteData;
+import org.apache.rocketmq.connect.hudi.config.*;
+import org.apache.rocketmq.remoting.RPCHook;
+import org.apache.rocketmq.tools.admin.DefaultMQAdminExt;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
```
Comment (ltamber) [inline rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java:37]: import * is not recommended

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    public static String HUDI_CONNECTOR_ADMIN_PREFIX = "HUDI-CONNECTOR-ADMIN";
+    public static final String PREFIX = "hudi";
+
+    public HudiSinkConnector() {
+        topicRouteMap = new HashMap<>();
+        sinkConnectConfig = new SinkConnectConfig();
+        executor = Executors.newSingleThreadScheduledExecutor(new BasicThreadFactory.Builder().namingPattern("HudiFSinkConnector-SinkWatcher-%d").daemon(true).build());
+    }
+
+    private synchronized void startMQAdminTools() {
```
Comment (odbozhou) [inline rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java:65]: What is the function of the startMQAdminTools method?

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        return "";
+    }
+
+    @Override
+    public void start() {
+        startMQAdminTools();
+        startListener();
+    }
+
+    public void startListener() {
```
Comment (odbozhou) [inline rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java:110]: What is the function of the startListener method?

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+            if (!originTasks.containsAll(updateTasks)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    public void buildRoute() {
```
Comment (odbozhou) [inline rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java:152]: What is the function of the buildRoute method?

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkTask.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+    public HudiSinkTask() {
+        this.hudiConnectConfig = new HudiConnectConfig();
+    }
+
+    @Override
+    public void put(Collection<SinkDataEntry> sinkDataEntries) {
+        try {
+            if (tableQueue.size() > 1) {
+                updater = tableQueue.poll(1000, TimeUnit.MILLISECONDS);
```
Comment (odbozhou) [inline rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkTask.java:57]: Sink starts with only one updater, why should a blocking queue be introduced?

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkTask.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+
+    }
+
+    @Override
+    public void stop() {
+        try {
+            for(Updater updater : tableQueue) {
+                updater.start();
```
Comment (odbozhou) [inline rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkTask.java:107]: start ?

File: rocketmq-connect-hudi/src/main/java/example/avro/User.java [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+ */
+
+import org.apache.avro.message.BinaryMessageDecoder;
+import org.apache.avro.message.BinaryMessageEncoder;
+import org.apache.avro.message.SchemaStore;
+import org.apache.avro.specific.SpecificData;
+import org.apache.avro.util.Utf8;
+
+@org.apache.avro.specific.AvroGenerated
+public class User extends org.apache.avro.specific.SpecificRecordBase implements org.apache.avro.specific.SpecificRecord {
```
Comment (odbozhou) [inline rocketmq-connect-hudi/src/main/java/example/avro/User.java:14]: Is this file necessary? If it is a necessary class, please modify the package name and include the apache license

**Review by odbozhou — COMMENTED** [review]
Pom can add maven-checkstyle-plugin to unify the code style

**Comment by odbozhou** [comment]
Submit an issue, this pr is associated with this issue

**Comment by lizhiboo** [comment]
#801

**Review by duhenglucky — COMMENTED** [review]
Quick start and config docs are needed.

File: rocketmq-connect/rocketmq-connect-runtime/src/main/java/org/apache/rocketmq/connect/runtime/connectorwrapper/WorkerSinkTask.java [hunk 9]
```diff
@@ -21,15 +21,10 @@
...
 import io.openmessaging.connector.api.PositionStorageReader;
 import io.openmessaging.connector.api.common.QueueMetaData;
-import io.openmessaging.connector.api.data.Converter;
-import io.openmessaging.connector.api.data.DataEntryBuilder;
-import io.openmessaging.connector.api.data.EntryType;
-import io.openmessaging.connector.api.data.Field;
-import io.openmessaging.connector.api.data.Schema;
-import io.openmessaging.connector.api.data.SinkDataEntry;
-import io.openmessaging.connector.api.data.SourceDataEntry;
+import io.openmessaging.connector.api.data.*;
```
Comment (duhenglucky) [inline rocketmq-connect/rocketmq-connect-runtime/src/main/java/org/apache/rocketmq/connect/runtime/connectorwrapper/WorkerSinkTask.java:24]: It is not recommended to use * for imports

**Review by duhenglucky — APPROVED** [review]

File: rocketmq-connect/rocketmq-connect-runtime/src/main/resources/connect.conf [hunk 10]
```diff
@@ -13,11 +13,16 @@
...
 ## Http port for user to access REST API
-httpPort=8081
+httpPort=8082
 
 # Rocketmq namesrvAddr
 namesrvAddr=localhost:9876
 
 # Source or sink connector jar file dir,The default value is rocketmq-connect-sample
-pluginPaths=/usr/local/connector-plugins
+pluginPaths=/Users/osgoo/Downloads/connector-plugins
```
Comment (odbozhou) [inline rocketmq-connect/rocketmq-connect-runtime/src/main/resources/connect.conf:26]: Personal directory ？

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+                        for (int i = 0; i < qd.getReadQueueNums(); i++) {
+                            MessageQueue taskTopicInfo = new MessageQueue(topic, qd.getBrokerName(), i);
+                            topicRouteMap.get(topic).add(taskTopicInfo);
+                        }
+                    }
+                }
+            }
+        } catch (Exception e) {
+            log.error("Fetch topic list error.", e);
+        } finally {
```
Comment (odbozhou) [inline rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java:190]: if finally unuse ，please remove

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+
+
+    /**
+     * We need to reason why we don't call srcMQAdminExt.shutdown() here, and why
+     * it can be applied to srcMQAdminExt
+     */
+    @Override
+    public void stop() {
+        listenerHandle.cancel(true);
+        // srcMQAdminExt.shutdown();
```
Comment (odbozhou) [inline rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java:203]: remove unused code

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/strategy/ITaskDivideStrategy.java [hunk 13]
```diff
@@ -0,0 +1,13 @@
+package org.apache.rocketmq.connect.hudi.strategy;
+
+import io.openmessaging.KeyValue;
+
+import java.util.List;
+
+/**
+ * @author osgoo
```
Comment (odbozhou) [inline rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/strategy/ITaskDivideStrategy.java:8]: please remove author info

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/strategy/TaskDivideByQueueStrategy.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+import io.openmessaging.internal.DefaultKeyValue;
+import org.apache.rocketmq.common.message.MessageQueue;
+import org.apache.rocketmq.connect.hudi.config.HudiConnectConfig;
+
+import java.util.*;
+
+import static org.apache.rocketmq.connect.hudi.config.HudiConnectConfig.*;
+
+/**
+ * @author osgoo
```
Comment (odbozhou) [inline rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/strategy/TaskDivideByQueueStrategy.java:14]: please remove author info

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/strategy/TaskDivideStrategyFactory.java [hunk 15]
```diff
@@ -0,0 +1,11 @@
+package org.apache.rocketmq.connect.hudi.strategy;
+
+/**
+ * @author osgoo
```
Comment (odbozhou) [inline rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/strategy/TaskDivideStrategyFactory.java:4]: please remove author info

File: rocketmq-connect/rocketmq-connect-runtime/src/main/java/org/apache/rocketmq/connect/runtime/ConnectStartup.java [hunk 16]
```diff
@@ -48,6 +48,7 @@
 
     public static void main(String[] args) {
 
+        args = new String[] {"-c", "/Users/osgoo/Documents/workspace/lizhiboo/rocketmq-externals/rocketmq-connect/rocketmq-connect-runtime/target/distribution/conf/connect.conf"};
```
Comment (odbozhou) [inline rocketmq-connect/rocketmq-connect-runtime/src/main/java/org/apache/rocketmq/connect/runtime/ConnectStartup.java:51]: please remove test code

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/config/CloneUtils.java (hunk 1 above)
Comment (lizhiboo) [reply rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/config/CloneUtils.java]: done

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java (hunk 2 above)
Comment (lizhiboo) [reply rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java]: done

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkTask.java (hunk 6 above)
Comment (lizhiboo) [reply rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkTask.java]: done

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkTask.java (hunk 7 above)
Comment (lizhiboo) [reply rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkTask.java]: done

File: rocketmq-connect-hudi/src/main/java/example/avro/User.java (hunk 8 above)
Comment (lizhiboo) [reply rocketmq-connect-hudi/src/main/java/example/avro/User.java]: removed

**Comment by lizhiboo** [comment]
> Pom can add maven-checkstyle-plugin to unify the code style

added

File: rocketmq-connect/rocketmq-connect-runtime/src/main/resources/connect.conf (hunk 10 above)
Comment (lizhiboo) [reply rocketmq-connect/rocketmq-connect-runtime/src/main/resources/connect.conf]: changed

File: rocketmq-connect/rocketmq-connect-runtime/src/main/java/org/apache/rocketmq/connect/runtime/connectorwrapper/WorkerSinkTask.java (hunk 9 above)
Comment (lizhiboo) [reply rocketmq-connect/rocketmq-connect-runtime/src/main/java/org/apache/rocketmq/connect/runtime/connectorwrapper/WorkerSinkTask.java]: done

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java (hunk 11 above)
Comment (lizhiboo) [reply rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java:192]: changed

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java (hunk 12 above)
Comment (lizhiboo) [reply rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/connector/HudiSinkConnector.java]: done

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/strategy/ITaskDivideStrategy.java (hunk 13 above)
Comment (lizhiboo) [reply rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/strategy/ITaskDivideStrategy.java]: done

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/strategy/TaskDivideByQueueStrategy.java (hunk 14 above)
Comment (lizhiboo) [reply rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/strategy/TaskDivideByQueueStrategy.java]: done

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/strategy/TaskDivideStrategyFactory.java (hunk 15 above)
Comment (lizhiboo) [reply rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/strategy/TaskDivideStrategyFactory.java]: done

File: rocketmq-connect/rocketmq-connect-runtime/src/main/java/org/apache/rocketmq/connect/runtime/ConnectStartup.java (hunk 16 above)
Comment (lizhiboo) [reply rocketmq-connect/rocketmq-connect-runtime/src/main/java/org/apache/rocketmq/connect/runtime/ConnectStartup.java]: done

File: rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/config/CloneUtils.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+            ObjectOutputStream oos = new ObjectOutputStream(baos);
+            oos.writeObject(obj);
+            oos.close();
+
+            ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray());
+            ObjectInputStream ois = new ObjectInputStream(bais);
+            clonedObj = (T) ois.readObject();
+            ois.close();
+        } catch (Exception e) {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (lizhiboo) [reply rocketmq-connect-hudi/src/main/java/org/apache/rocketmq/connect/hudi/config/CloneUtils.java:41]: done

**Review by duhenglucky — APPROVED** [review]
